Ovsiannikov78/Angular-and-Spring-Boot
# 07_02_201 eCommerce Project - Angular Front End - Product List

### Angular Front End 

**1. Product List**

- Set up;
- Create Product class and Product Service;
- Update Angular Component;
- Update Angular Component and Integrating Angular and Spring Boot for Full Stack App!
   Added CrossOrigin annotation to the backend part of project;
- Building an HTML Table;
- Adding Product Images

**2. Integrating Online Shop Template**

- Install Bootstrap CSS Styles with npm;
- Add local custom CSS styles;
- Integrate Template Files and Logos;
- Add support for Product Grid;
- Enhance Application with Product Images

**3. Search for Products by Category** 

- Define routes;
- Define Router Outlet;
- Set up Router Links to pass category id param;
- Enhance ProductListComponent to read category id param;
- Update Angular Service to call new URL on Spring Boot app

**4. Search for Products by Category (Dynamic Search Component)**

- Create a class - ProductCategory;
- Create new component for menu;
- Enhance menu component to read data from product service;
- Update product service to call URL on Spring Boot app;
- In HTML, replace hard-coded links with menu component

**5. Display Category Name in Product List Grid**

- Update route and add parameter;
- Update routeLink to include the category name;
- Update component to read the category name;
- Update the list-grid HTML file to bind to category name

**6. Search for Products by Keyword**

- Create new component fro search;
- Add new Angular route for searching;
- Update SearchComponent to send data to search route;
- Enhance ProductListComponent to search for products with ProductService;
- Update ProductService to call URL on Spring Boot app;
- Enhancement - display a message if products not found

**7. Product Master-Detail View**

- Create new component for product details;
- Add new Angular route for product detail;
- Add router links to the product-list-grid HTML page;
- Enhance ProductDetailsComponent to retrieve product from ProductService;
- Update ProductService to call URL on Spring Boot app;
- Update HTML page for ProductDetailsComponent to display product details;
- Add new CSS style to the file: styles.css;
- Add more Product Details;
- Race Conditions - Solution: Create a new instance

**8. Pagination**

- Install ng-bootstrap;
- Refactor the interface for: GetResponseProduct;
- Add pagination support to ProductService;
- Update ProductListComponent to handle pagination;
- Enhance HTML template to use ng-bootstrap pagination component

**9. Functionality page size selection**

- Add drop-down list for page size to HTML template;
- Update ProductListComponent for setting page size;
- Max size;
- Boundary Links - Ability to jump from the first page immediately to the last  and back
- 

**10. Keyword Search**

- Add pagination support to ProductService;
- Update ProductListComponent to handle pagination

**11. Shopping Cart Status Component**

- Create new component: CartStatusComponent;
- Add HTML template for CartStatusComponent;
- Add click handler for "Add to cart" button;
- Update ProductListComponent with click handler method;
- Create model class: CartItem;
- Develop CartService;
- Modify ProductListComponent to call CartService;
- Enhance CartStatusComponent to subscribe to CartService;
- Update CartStatusComponent HTML to display cart total price and quantity;
- Refactor `cart.service.ts`;
- Add To Cart From Details View

**12. Shopping Cart Items CRUD**

- #### 1.  List Shopping Cart Items 

- Create new component: CartDetailsComponent;
- Add new router for CartDetailsComponent;
- Update link for Shopping Cart icon;
- Create HTML Table;
- Modify CartDetailsComponent to retrieve cart items;
- Add HTML template for CartDetailsComponent;
- Display Cart Totals;
- Handle an Empty Cart

- #### 2. Increment Item Quantity

- Modify CartDetailsComponent HTML template - Add the "increment" button;
- Modify CartDetailsComponent HTML template - Add click handler for the "increment" button on HTML template
- Update CartDetailsComponent with click handler method;

- #### 3. Decrement/Remove Item Quantity

- Add click handler for the "decrement" button on HTML template;
- Update CartDetailsComponent with click handler method;
- Modify CartDetailsComponent with supporting method;
- Repeat process for "Remove" button

**13. Checkout Form Layout**

- Generate our checkout component;
- Add a new route for checkout component;
- Create a new checkout button and link to checkout component;
- Add support for reactive forms;
- Define form in `component.ts` file;
- Layout form controls in HTML template;
- Add event handler for form submission;
- Shipping Address;
- Billing Address;
- Add check box to copy Shipping Address to Billing Address;
- Credit Card;
- Review Your Order

**14. Checkout Form - Drop-Down Lists**

- Generate form service: Luv2ShopFormService;
- Add methods to the form service for month and years;
- Update checkout component to retrieve the months and years from service;
- Update HTML template to populate drop-down lists for months and years;
- Dependent Fields:

#### 1. Update HTML template;
- For the Expiration Years drop-down list;
- Add event binding for change event;

#### 2. Update checkout component, add event handler;
- #### Read the selected year;
- #### Update the list of months based on selected year

- Create TypeScript classes for: Country and State;
- Add methods to the form service for countries and states;
- Update checkout component to retrieve the countries from service;
- Update HTML template to populate drop-down list for countries;
- Add event handler for checkout component;

- #### Read the selected country, retrieve list of states based on selected country;

- Update HTML template to populate drop-down list for states;
- Add console logs;
- Bug Fix

**15. Checkout Form - Validation**
- #### Display Validation Error Messages:

- Specify validation rules for the form controls;
- Define Getter methods to access form controls;
- Update HTML template to display error message;
- Add event handler to check validation status when submit button clicked;

- #### Custom Validator Rule:

- Define custom validator rule;
- Specify custom validator rule for the form controls;
- Update HTML template to display error messages;

- #### Shipping Validation:
- #### Billing Validation:
- #### Credit Card Validation

**16. Checkout Form - Review Cart Totals**
- #### 1. Updates for CheckoutComponent
  

- Inject CartService into CheckoutComponent
- In `ngOnit` method, call new method: `reviewCartDetails()`;
- Add code for new method: `reviewCartDetails()`;

- #### 2. Update for CartService
- Change Subject to BehaviorSubject

**17. Checkout Form - Save the Order to Database**
- #### 1. Create common classes
- Customer, Order, OrderItem, Address, Purchase

- #### 2. Create CheckoutService
- Make REST API call to Spring Boot backend
- Update onSubmit() method to collect form data, call CheckoutService

- #### 3. Update CheckoutComponent
- Inject CheckoutService and Router 

**18. Security - Login/Logout**

- Create the Login Component;
- Code Refactor - moving all components to a Components folder;
- Render the Sign-In Widget;
- Develop login status component for login/logout buttons;
- Add CSS for Login Status Component;

- #### 1. Update App Module configs to connect routes:
- Connecting Routes;
- App Module Configuration;

**19. Security - User Registration**

- Configure code in login component;


**20. VIP Member Access - Protected Routes**
- Generate members-page component;
- Update template text in HTML page;
- Add "Member" button to login-status component;
- Define protected route for members-page component;
- Adding order-history class, order-history component, order-history service
